book/src/pong-tutorial/pong-tutorial-04.md

```diff
@@ -0,0 +1,351 @@
+the elapsed time.
+As you can see, to gain access to time passed since the last frame, you need
+to use the [`amethyst::core::timing::Time`][doc_time], a commonly used
+resource. It has a method called `delta_seconds` that does exactly what we want.
+
+Then, we'll implement the `BounceSystem` in `systems/bounce.rs`:
+```rust,no_run,noplaypen
+# extern crate amethyst;
+# use amethyst::ecs::prelude::{Component, DenseVecStorage};
+# pub struct Ball {
+#    pub velocity: [f32; 2],
+#    pub radius: f32,
+# }
+# impl Component for Ball {
+#    type Storage = DenseVecStorage<Self>;
+# }
+# #[derive(PartialEq, Eq)]
+# pub enum Side {
+#   Left,
+#   Right,
+# }
+# pub struct Paddle {
+#   side: Side,
+#   width: f32,
+#   height: f32,
+# }
+# impl amethyst::ecs::Component for Paddle {
+#   type Storage = amethyst::ecs::VecStorage<Paddle>;
+# }
+# const PADDLE_HEIGHT: f32 = 16.0;
+# const PADDLE_WIDTH: f32 = 4.0;
+# const SPRITESHEET_SIZE: (f32, f32) = (8.0, 16.0);
+# const BALL_RADIUS: f32 = 2.0;
+# const BALL_VELOCITY_X: f32 = 75.0;
+# const BALL_VELOCITY_Y: f32 = 50.0;
+# const ARENA_HEIGHT: f32 = 100.0;
+# const ARENA_WIDTH: f32 = 100.0;
+use amethyst::{
+    core::transform::Transform,
+    ecs::prelude::{Join, ReadStorage, System, WriteStorage},
+};
+
+pub struct BounceSystem;
+
+impl<'s> System<'s> for BounceSystem {
+    type SystemData = (
+        WriteStorage<'s, Ball>,
+        ReadStorage<'s, Paddle>,
+        ReadStorage<'s, Transform>,
+    );
+
+    fn run(&mut self, (mut balls, paddles, transforms): Self::SystemData) {
+        // Check whether a ball collided, and bounce off accordingly.
+        //
+        // We also check for the velocity of the ball every time, to prevent multiple collisions
+        // from occurring.
+        for (ball, transform) in (&mut balls, &transforms).join() {
+            let ball_x = transform.translation[0];
+            let ball_y = transform.translation[1];
+
+            // Bounce at the top or the bottom of the arena.
+            if ball_y <= ball.radius && ball.velocity[1] < 0.0 {
+                ball.velocity[1] = -ball.velocity[1];
+            } else if ball_y >= ARENA_HEIGHT - ball.radius && ball.velocity[1] > 0.0 {
+                ball.velocity[1] = -ball.velocity[1];
+            }
+
+            // Bounce at the paddles.
+            for (paddle, paddle_transform) in (&paddles, &transforms).join() {
+                let paddle_x = paddle_transform.translation[0] - paddle.width * 0.5;
+                let paddle_y = paddle_transform.translation[1] - paddle.height * 0.5;
+
+                // To determine whether the ball has collided with a paddle, we create a larger
+                // rectangle around the current one, by subtracting the ball radius from the
+                // lowest coordinates, and adding the ball radius to the highest ones. The ball
+                // is then within the paddle if its centre is within the larger wrapper
+                // rectangle.
```

Comment:
If you've got time, maybe a picture could help here?
It's okay if you don't, I just want to make sure it's perfectly clear what we are doing.